addons/resource/models/resource.py

```diff
@@ -985,6 +985,50 @@ def _get_unavailable_intervals(self, start, end):
+        """
+            Returns a dict of dict with resource's id as first key and resource's calendar as secondary key
+            The value is the validity interval of the calendar for the given resource.
+
+            Here the validity interval for each calendar is the whole interval but it's meant to be overriden in further modules
+            handling resource's employee contracts.
+        """
+        assert start.tzinfo and end.tzinfo
+        calendars_within_period_per_resource = defaultdict(lambda: defaultdict(Intervals))  # keys are [resource id:integer][calendar:self.env['resource.calendar']]
+        default_calendar = default_company and default_company.resource_calendar_id or self.env.company.resource_calendar_id
+        if not self:
+            # if no resource, add the company resource calendar.
+            calendars_within_period_per_resource[False][default_calendar] = Intervals([(start, end, self.env['resource.calendar.attendance'])])
+        for resource in self:
+            calendar = resource.calendar_id or resource.company_id.resource_calendar_id or default_calendar
+            calendars_within_period_per_resource[resource.id][calendar] = Intervals([(start, end, self.env['resource.calendar.attendance'])])
+        return calendars_within_period_per_resource
+
+    def _get_work_intervals_batch(self, start, end):
+        """
+            Returns the work intervals of the resource following their calendars between ``start`` and ``end``
+
+            This methods handle the eventuality of an resource having multiple resource calendars, see _get_calendars_validity_within_period method
+            for further explanation.
+        """
+        assert start.tzinfo and end.tzinfo
+        resource_calendar_validity_intervals = {}
+        resource_per_calendar = defaultdict(lambda: self.env['resource.resource'])
+        work_intervals_per_resource = defaultdict(Intervals)
+
+        resource_calendar_validity_intervals = self._get_calendars_validity_within_period(start, end)
+        for resource_id in self:
+            # For each resource, retrieve its calendar and their validity intervals
+            for calendar in resource_calendar_validity_intervals[resource_id.id].keys():
+                resource_per_calendar[calendar] |= resource_id
+        for calendar in resource_per_calendar.keys():
+            # For each calendar used by the resources, retrieve the work intervals for every resources using it
+            work_intervals_batch = calendar._work_intervals_batch(start, end, resources=resource_per_calendar[calendar])
+            for resource_id in resource_per_calendar[calendar]:
+                # Make the conjonction between work intervals and calendar validity
+                work_intervals_per_resource[resource_id.id] |= work_intervals_batch[resource_id.id] & resource_calendar_validity_intervals[resource_id.id][calendar]
+
+        return work_intervals_per_resource
```

Comment:
Hello 👋 

~~Since we are relying quite a bit already on `Intervals` I think we could go even further
I believe we could gather the leave intervals beforehand and subtract them from the work intervals we compute with validity periods on line 1029. Thus we could also use `_attendance_intervals_batch` instead of `_work_intervals_batch`.~~

After checking the code a bit more I realised this might not be possible and/or beneficial.
At first I thought personal leaves had no calendar linked but it is not the case (see https://github.com/odoo/odoo/blob/master/addons/hr_holidays/models/hr_leave.py#L961).

Appart from that there is another issue, global leaves, they are linked to the calendars, so we would have needed to call `_leave_intervals_batch` on every calendar anyway (which is done in `_work_intervals_batch`).

I think in this state there is not much that can be done to reduce the number of queries 🙂